src/org/stringtemplate/v4/STGroup.java

```diff
@@ -80,9 +80,11 @@
 		TEMPLATE_FILE_EXTENSION = ".st";
 	}
 
-    /** When we use key as a value in a dictionary, this is how we signify. */
-    public static final String DICT_KEY = "key";
-    public static final String DEFAULT_KEY = "default";
+    /** When we use key as a value in a dictionary, this is how we signify.
+     *  Magic suffix is used to distinguish "default" string literal and "default" key
+     *  End part is the date of this suffix adding. */
+    public static final String DICT_KEY = "key_st4_magic_suffix_2016/01/10";
+    public static final String DEFAULT_KEY = "default_st4_magic_suffix_2016/01/10";
```

Comment:
What about this test case:
```java

public static final String DEFAULT_KEY = "\0"; // If replace with "asdf", for example, test will be passed.

...

@Test public void testDictDefaultStringAsKey() throws Exception {
    String templates =
            "typeInit ::= [\"\0\":\"foo\", default:key] "+newline+
            "var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
            ;
    writeFile(tmpdir, "test.stg", templates);
    STGroup group = new STGroupFile(tmpdir+"/"+"test.stg");
    ST st = group.getInstanceOf("var");
    st.add("type", "\0");
    st.add("name", "x");
    String expecting = "\0 x = foo;";
    String result = st.render();
    assertEquals(expecting, result); // expecting: " x = foo;" result: " x = ;"
}
```
Of course it is unlikely but possible. Actually key can be any string due to escaping.